include/cuco/detail/trie/bit_vector/bit_vector.cuh

```diff
@@ -0,0 +1,200 @@
+template <class Extent             = cuco::experimental::extent<std::size_t>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          class Allocator          = cuco::cuda_allocator<std::byte>,
+          class Storage            = cuco::experimental::storage<1>>
+class bit_vector {
+ public:
+  bit_vector();
+  ~bit_vector();
+
+  /**
+   * @brief adds a new bit at the end
+   *
+   * Grows internal storage if needed
+   *
+   * @param bit Boolean value of new bit to be added
+   */
+  void append(bool bit);
+
+  /**
+   * @brief Builds indexes for rank and select
+   *
+   * Also creates device-side snapshot
+   */
+  void build();
+
+  using size_type = typename Extent::value_type;  ///< size type to specify bit index
+  /**
+   * @brief Modifies a single bit
+   *
+   * @param index position of bit to be modified
+   * @param bit new value of bit
+   */
+  void set(size_type index, bool bit);
+
+  /**
+   * @brief Sets last bit to specified value
+   *
+   * @param bit new value of last bit
+   */
+  void set_last(bool bit);
+
+  static constexpr auto cg_size      = 1;      ///< CG size used for probing
+  static constexpr auto window_size  = 1;      ///< Window size used for probing
+  static constexpr auto thread_scope = Scope;  ///< CUDA thread scope
+
+  using extent_type =
+    decltype(make_window_extent<cg_size, window_size>(std::declval<Extent>()));  ///< Extent type
+  using allocator_type = Allocator;                                              ///< Allocator type
+  using storage_type =
+    detail::storage<Storage, size_type, extent_type, allocator_type>;  ///< Storage type
+
+  using storage_ref_type = typename storage_type::ref_type;  ///< Non-owning window storage ref type
+  template <typename... Operators>
+  using ref_type =
+    cuco::experimental::bit_vector_ref<storage_ref_type,
+                                       Operators...>;  ///< Non-owning container ref type
+
+  /**
+   * @brief Get device ref with operators.
+   *
+   * @tparam Operators Set of `cuco::op` to be provided by the ref
+   *
+   * @param ops List of operators, e.g., `cuco::bv_read`
+   *
+   * @return Device ref of the current `bit_vector` object
+   */
+  template <typename... Operators>
+  [[nodiscard]] auto ref(Operators... ops) const noexcept;
+
+  /**
+   * @brief Get the number of bits bit_vector holds
+   *
+   * @return Number of bits bit_vector holds
+   */
+  size_t size() const { return n_bits_; }
+
+ private:
+  uint64_t n_bits_;  ///< Number of bits added to bit_vector
+
+  // Host-side structures
+  std::vector<uint64_t> words_;     ///< Words vector that represents all bits
+  std::vector<rank> ranks_;         ///< Holds the rank values for every 256-th bit (4-th word)
+  std::vector<rank> ranks0_;        ///< Same as ranks_ but for `0` bits
+  std::vector<uint64_t> selects_;   ///< Holds pointers to (0, 256, 512...)th `1` bit in ranks_
+  std::vector<uint64_t> selects0_;  ///< Same as selects_, but for `0` bits
```

Comment:
```suggestion
  std::vector<slot_type> words_;     ///< Words vector that represents all bits
  std::vector<rank> ranks_;         ///< Holds the rank values for every 256-th bit (4-th word)
  std::vector<rank> ranks0_;        ///< Same as ranks_ but for `0` bits
  std::vector<size_type> selects_;   ///< Holds indices to (0, 256, 512...)th `1` bit in ranks_
  std::vector<size_type> selects0_;  ///< Same as selects_, but for `0` bits
```
I could be wrong, but the idea here is to use `size_type` and `slot_type` consistently to avoid plain `uint64_t`.

Reply:
91b628a5925bec0b4a4c7f21dc6ebad40c8ba0d9